### DISCUSSION TITLE
new formula pg_cron cant install shared lib - permission denied

### DISCUSSION THREAD

**Discussion by mtesch-um** [Homebrew/discussions#3312] [Writing Formulae/Casks]
I'm trying to create a new formula for a very simple extension to postgresql - this extension simply builds a few c files into a shrared library `pg_cron.so` and installs that with some support files into `/usr/local/lib/postgresql/`.  The install of the shared object always fails.

Before running the install the path /usr/local/lib/postgresql` is a symlink into postgresql's Cellar, but after it's a directory with almost all the contents symlinked there.

Any advice for what I'm doing wrong here?  Here's the formula: https://github.com/mtesch-um/homebrew-core/tree/formula-pg_cron
 
Is this the best way to approach this?  There are several similarly-built interesting extensions to postgres that would be nice to have (pg_partman, citus).

Here's the error:
```
% brew reinstall --build-from-source pg_cron 
...
clang -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Werror=vla -Werror=unguarded-availability-new -Wendif-labels -Wmissing-format-attribute -Wcast-function-type -Wformat-security -fno-strict-aliasing -fwrapv -Wno-unused-command-line-argument -Wno-compound-token-split-by-macro -O2  -std=c99 -Wall -Wextra -Werror -Wno-unused-parameter -Wno-implicit-fallthrough -Iinclude -I/usr/local/include -I. -I./ -I/usr/local/include/postgresql/server -I/usr/local/include/postgresql/internal -I/usr/local/Cellar/icu4c/70.1/include -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/readline/include   -c -o src/task_states.o src/task_states.c
cat pg_cron.sql > pg_cron--1.0.sql
/bin/sh /usr/local/lib/postgresql/pgxs/src/makefiles/../../config/install-sh -c -d '/usr/local/lib/postgresql'
/bin/sh /usr/local/lib/postgresql/pgxs/src/makefiles/../../config/install-sh -c -d '/usr/local/share/postgresql/extension'
/bin/sh /usr/local/lib/postgresql/pgxs/src/makefiles/../../config/install-sh -c -d '/usr/local/share/postgresql/extension'
clang -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Werror=vla -Werror=unguarded-availability-new -Wendif-labels -Wmissing-format-attribute -Wcast-function-type -Wformat-security -fno-strict-aliasing -fwrapv -Wno-unused-command-line-argument -Wno-compound-token-split-by-macro -O2  -bundle -multiply_defined suppress -o pg_cron.so src/entry.o src/job_metadata.o src/misc.o src/pg_cron.o src/task_states.o -L/usr/local/lib  -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk -L/usr/local/opt/openssl@1.1/lib -L/usr/local/opt/readline/lib  -Wl,-dead_strip_dylibs   -L/usr/local/lib -lpq -bundle_loader /usr/local/Cellar/postgresql/14.3/bin/postgres
/usr/bin/install -c -m 755  pg_cron.so '/usr/local/lib/postgresql/pg_cron.so'
install: /usr/local/lib/postgresql/pg_cron.so: Operation not permitted
```

**Comment by carlocab** [comment]
You can’t install directly into `/usr/local`; that location is protected by the sandbox. You should install into the formula’s `prefix`. Most other formulae in homebrew-core should show examples of this.

**Reply by SMillerDev** [reply]
Additionally, homebrew won't let you install into the prefix of another formula. You'll have to find something like `/usr/local/share` to install to

**Reply by mtesch-um** [reply]
Thanks for the quick response!

Ultimately what I think *should* result is only symlinks in `/usr/local/lib/postgresql/` - some go to the postgresql formula's Cellar, and some go to the pg_cron formula's cellar - is this reasonable?  It's postgres' default module load path.

I think I was able to fix it in the latest commit - not sure if this is the best way though?  Tests pass, so I'm making a PR as well.

**Reply by carlocab** [reply]
> Ultimately what I think _should_ result is only symlinks in `/usr/local/lib/postgresql/` - some go to the postgresql formula's Cellar, and some go to the pg_cron formula's cellar - is this reasonable?

No, since you can’t install into another formula’s prefix. You might wish to look at other formulae that install Postgres extensions to see how they handle this.

**Reply by mtesch-um** [reply]
The new version seems to be working the way I would expect, and passes tests - both formulas now share `/usr/local/lib/postgresql/` (at least on my machine.)  Is there a problem like this?  Sorry I dont understand brew well enough, is `/usr/local/lib/postgresql/` postgresql's prefix?  is the prefix it's Cellar dir, or `/usr/local/`?

Thanks for the tip about other postgres extensions - didn't realize there were others - even citus(!).  The ones I've looked at so far use `configure`.

**Reply by mtesch-um** [reply]
PR: https://github.com/Homebrew/homebrew-core/pull/102345
